pkg/sql/opt/exec/execbuilder/tests/local/local_test: TestExecBuild_partial_stats failed

pkg/sql/opt/exec/execbuilder/tests/local/local_test.TestExecBuild_partial_stats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16034799?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16034799?buildTab=artifacts#/) on master @ [9ed67ddc4a3f09361ce680327f38501ef0b9167f](https://github.com/cockroachdb/cockroach/commits/9ed67ddc4a3f09361ce680327f38501ef0b9167f):


```
=== RUN   TestExecBuild_partial_stats
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/e99f90515aafe7a991d00d2dcb22e05e/logTestExecBuild_partial_stats3784852338
    test_log_scope.go:81: use -show-logs to present logs inline
    logic.go:3014: let $forecastsEnabledPrev = false
    logic.go:3014: let $stat_id_simple = 985192213729476609
    logic.go:3014: let $stat_id_empty = 985192213884796929
    logic.go:3014: let $stat_id_start_and_end = 985192213976055809
    logic.go:3014: let $stat_id_full_a = 985192214096314369
    logic.go:3014: let $stat_id_full_b = 985192214088646657
    logic.go:3014: let $hist_id_ka_outer_buckets_full = 985192215920508929
    logic.go:2968: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/10327/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1391: EXPLAIN SELECT * FROM ka WHERE a > 5
        expected:
            distribution: local
            vectorized: true
            ·
            • scan
              estimated row count: 4 (36% of the table; stats collected <hidden> ago)
              table: ka@ka_a_idx
              spans: [/6 - ]
            
        but found (query options: "") :
            distribution: local
            vectorized: true
            ·
            • scan
              estimated row count: 4 (40% of the table; stats collected <hidden> ago)
              table: ka@ka_a_idx
              spans: [/6 - ]
            
[19:46:45] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/10327/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats with config local: 55 tests, 1 failures
    logic.go:4135: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/10327/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1402: error while processing
    logic.go:4135: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/10327/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1402: too many errors encountered, skipping the rest of the input
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/e99f90515aafe7a991d00d2dcb22e05e/logTestExecBuild_partial_stats3784852338
--- FAIL: TestExecBuild_partial_stats (2.21s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExecBuild_partial_stats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-40242

pkg/sql/opt/exec/execbuilder/tests/local/local_test.TestExecBuild_partial_stats [failed](https://tanzanite.cluster.engflow.com/invocations/default/879e9bb2-8799-48b5-904e-34bfd1d30837?testReportRun=17&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvc3FsL29wdC9leGVjL2V4ZWNidWlsZGVyL3Rlc3RzL2xvY2FsOmxvY2FsX3Rlc3Q=) on master @ [bcd4e5e37f2a95e73347f5ed83db5a24bebfd8d1](https://github.com/cockroachdb/cockroach/commits/bcd4e5e37f2a95e73347f5ed83db5a24bebfd8d1):


```
=== RUN   TestExecBuild_partial_stats
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestExecBuild_partial_stats1221029553
    test_log_scope.go:81: use -show-logs to present logs inline
    logic.go:3014: let $forecastsEnabledPrev = false
    logic.go:3014: let $stat_id_simple = 986740916074151937
    logic.go:3014: let $stat_id_empty = 986740916173897729
    logic.go:3014: let $stat_id_start_and_end = 986740916245495809
    logic.go:3014: let $stat_id_full_a = 986740916331905025
    logic.go:3014: let $stat_id_full_b = 986740916328005633
    logic.go:3014: let $hist_id_ka_outer_buckets_full = 986740920296243201
[07:03:52] --- progress: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats: 50 statements
    logic.go:2968: 
         
        /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1391: EXPLAIN SELECT * FROM ka WHERE a > 5
        expected:
            distribution: local
            vectorized: true
            ·
            • scan
              estimated row count: 4 (36% of the table; stats collected <hidden> ago)
              table: ka@ka_a_idx
              spans: [/6 - ]
            
        but found (query options: "") :
            distribution: local
            vectorized: true
            ·
            • scan
              estimated row count: 4 (40% of the table; stats collected <hidden> ago)
              table: ka@ka_a_idx
              spans: [/6 - ]
            
[07:03:52] --- done: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats with config local: 55 tests, 1 failures
    logic.go:4135: 
        /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1402: error while processing
    logic.go:4135: /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1402: too many errors encountered, skipping the rest of the input
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestExecBuild_partial_stats1221029553
--- FAIL: TestExecBuild_partial_stats (2.96s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=17</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExecBuild_partial_stats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

pkg/sql/opt/exec/execbuilder/tests/local/local_test.TestExecBuild_partial_stats [failed](https://tanzanite.cluster.engflow.com/invocations/default/d98c7ab9-ab5f-42e9-8e2f-354d9fed7730?testReportRun=12&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvc3FsL29wdC9leGVjL2V4ZWNidWlsZGVyL3Rlc3RzL2xvY2FsOmxvY2FsX3Rlc3Q=) on master @ [4af6b13bf95a52fe8f6857ffd32e8c1156c36950](https://github.com/cockroachdb/cockroach/commits/4af6b13bf95a52fe8f6857ffd32e8c1156c36950):


```
=== RUN   TestExecBuild_partial_stats
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestExecBuild_partial_stats3455901797
    test_log_scope.go:81: use -show-logs to present logs inline
    logic.go:3014: let $forecastsEnabledPrev = false
    logic.go:3014: let $hist_id_ka_outer_buckets_full = 988712991594708993
    logic.go:2968: 
         
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1365: EXPLAIN SELECT * FROM ka WHERE a > 5
        expected:
            distribution: local
            vectorized: true
            ·
            • scan
              estimated row count: 4 (36% of the table; stats collected <hidden> ago)
              table: ka@ka_a_idx
              spans: [/6 - ]
            
        but found (query options: "") :
            distribution: local
            vectorized: true
            ·
            • scan
              estimated row count: 4 (40% of the table; stats collected <hidden> ago)
              table: ka@ka_a_idx
              spans: [/6 - ]
            
[06:14:20] --- done: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats with config local: 50 tests, 1 failures
    logic.go:4135: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1376: error while processing
    logic.go:4135: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1376: too many errors encountered, skipping the rest of the input
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestExecBuild_partial_stats3455901797
--- FAIL: TestExecBuild_partial_stats (2.13s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=12</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExecBuild_partial_stats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

pkg/sql/opt/exec/execbuilder/tests/local/local_test.TestExecBuild_partial_stats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16207569?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16207569?buildTab=artifacts#/) on master @ [a9b5972130abc204ef68117adbf6f402e55706e6](https://github.com/cockroachdb/cockroach/commits/a9b5972130abc204ef68117adbf6f402e55706e6):


```
=== RUN   TestExecBuild_partial_stats
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/a0740b839b2d4f82096b3077c77913d1/logTestExecBuild_partial_stats886897388
    test_log_scope.go:81: use -show-logs to present logs inline
    logic.go:3014: let $forecastsEnabledPrev = false
    logic.go:3014: let $hist_id_ka_outer_buckets_full = 988895949967884289
    logic.go:2968: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9129/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1365: EXPLAIN SELECT * FROM ka WHERE a > 5
        expected:
            distribution: local
            vectorized: true
            ·
            • scan
              estimated row count: 4 (36% of the table; stats collected <hidden> ago)
              table: ka@ka_a_idx
              spans: [/6 - ]
            
        but found (query options: "") :
            distribution: local
            vectorized: true
            ·
            • scan
              estimated row count: 4 (40% of the table; stats collected <hidden> ago)
              table: ka@ka_a_idx
              spans: [/6 - ]
            
[21:44:54] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9129/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats with config local: 50 tests, 1 failures
    logic.go:4135: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9129/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1376: error while processing
    logic.go:4135: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9129/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/partial_stats:1376: too many errors encountered, skipping the rest of the input
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a0740b839b2d4f82096b3077c77913d1/logTestExecBuild_partial_stats886897388
--- FAIL: TestExecBuild_partial_stats (2.85s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExecBuild_partial_stats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/distsql_stats
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_stats
@@ -2410,15 +2410,10 @@ INSERT INTO a_null VALUES (NULL), (1), (2);
 statement ok
 CREATE STATISTICS a_null_stat ON a FROM a_null;
 
-let $a_null_stats
-SHOW STATISTICS USING JSON FOR TABLE a_null
-
-# Inject statistics so that the current stats
-# cache is invalidated and creating partial
-# statistics has access to the latest full
-# statistic.
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
 statement ok
-ALTER TABLE a_null INJECT STATISTICS '$a_null_stats'
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement ok
 INSERT INTO a_null VALUES (NULL), (NULL), (NULL);
@@ -2453,15 +2448,10 @@ INSERT INTO d_desc VALUES (1, 10), (2, 20), (3, 30), (4, 40);
 statement ok
 CREATE STATISTICS sd ON a FROM d_desc;
 
-let $d_desc_stats
-SHOW STATISTICS USING JSON FOR TABLE d_desc;
-
-# Inject statistics so that the current stats
-# cache is invalidated and creating partial
-# statistics has access to the latest full
-# statistic.
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
 statement ok
-ALTER TABLE d_desc INJECT STATISTICS '$d_desc_stats'
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement ok
 INSERT INTO d_desc VALUES (0, 0), (5, 50);
@@ -2495,15 +2485,10 @@ INSERT INTO d_desc VALUES (NULL, NULL), (NULL, 2);
 statement ok
 CREATE STATISTICS sdn ON a FROM d_desc;
 
-let $d_desc_null_stats
-SHOW STATISTICS USING JSON FOR TABLE d_desc;
-
-# Inject statistics so that the current stats
-# cache is invalidated and creating partial
-# statistics has access to the latest full
-# statistic.
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
 statement ok
-ALTER TABLE d_desc INJECT STATISTICS '$d_desc_null_stats'
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement ok
 CREATE STATISTICS sdnp ON a FROM d_desc USING EXTREMES;
@@ -2535,18 +2520,13 @@ INSERT INTO s VALUES ('c'), ('d'), ('e');
 statement ok
 CREATE STATISTICS str ON s FROM s;
 
-let $s_stats
-SHOW STATISTICS USING JSON FOR TABLE s;
-
 statement ok
 INSERT INTO s VALUES ('a'), ('b'), ('f');
 
-# Inject statistics so that the current stats
-# cache is invalidated and creating partial
-# statistics has access to the latest full
-# statistic.
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
 statement ok
-ALTER TABLE s INJECT STATISTICS '$s_stats'
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement ok
 CREATE STATISTICS s_partial ON s FROM s USING EXTREMES;
@@ -2570,15 +2550,10 @@ INSERT INTO j VALUES ('{"1":"10"}'), ('{"2":"20"}'),  ('{"3":"30"}');
 statement ok
 CREATE STATISTICS j_full ON j FROM j;
 
-let $j_stats
-SHOW STATISTICS USING JSON FOR TABLE j;
-
-# Inject statistics so that the current stats
-# cache is invalidated and creating partial
-# statistics has access to the latest full
-# statistic.
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
 statement ok
-ALTER TABLE j INJECT STATISTICS '$j_stats'
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement error pq: table j does not contain a non-partial forward index with j as a prefix column
 CREATE STATISTICS j_partial ON j FROM j USING EXTREMES;
@@ -2604,15 +2579,10 @@ INSERT INTO only_null VALUES (NULL), (NULL), (NULL);
 statement ok
 CREATE STATISTICS only_null_stat ON a FROM only_null;
 
-let $only_null_stat
-SHOW STATISTICS USING JSON FOR TABLE only_null;
-
-# Inject statistics so that the current stats
-# cache is invalidated and creating partial
-# statistics has access to the latest full
-# statistic.
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
 statement ok
-ALTER TABLE only_null INJECT STATISTICS '$only_null_stat';
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement error pq: only outer or NULL bounded buckets exist in the index, so partial stats cannot be collected
 CREATE STATISTICS only_null_partial ON a FROM only_null USING EXTREMES;
@@ -2953,6 +2923,9 @@ CREATE TABLE int_outer_buckets (a PRIMARY KEY) AS SELECT generate_series(0, 9999
 statement ok
 CREATE STATISTICS int_outer_buckets_full ON a FROM int_outer_buckets;
 
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
+
 let $hist_id_int_outer_buckets_full
 SELECT histogram_id FROM [SHOW STATISTICS FOR TABLE int_outer_buckets] WHERE statistics_name = 'int_outer_buckets_full'
 
@@ -2990,6 +2963,11 @@ SET CLUSTER SETTING sql.stats.histogram_samples.count = 10050;
 statement ok
 CREATE STATISTICS int_outer_buckets_full ON a FROM int_outer_buckets;
 
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
+
 statement ok
 CREATE STATISTICS int_outer_buckets_partial ON a FROM int_outer_buckets USING EXTREMES;
 
@@ -3017,6 +2995,11 @@ INSERT INTO timestamp_outer_buckets VALUES
 
 statement ok
 CREATE STATISTICS timestamp_outer_buckets_full ON a FROM timestamp_outer_buckets;
+
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
 
 let $hist_id_timestamp_outer_buckets_full
 SELECT histogram_id FROM [SHOW STATISTICS FOR TABLE timestamp_outer_buckets] WHERE statistics_name = 'timestamp_outer_buckets_full'
@@ -3128,6 +3111,11 @@ INSERT INTO bool_table VALUES (true), (false)
 statement ok
 CREATE STATISTICS bool_table_full ON a FROM bool_table
 
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
+
 statement error pgcode 0A000 creating partial statistics at extremes on bool and enum columns is disabled
 CREATE STATISTICS bool_table_partial ON a FROM bool_table USING EXTREMES;
 
@@ -3139,6 +3127,11 @@ INSERT INTO enum_table VALUES ('hello'), ('howdy'), ('hi')
 
 statement ok
 CREATE STATISTICS enum_table_full ON a FROM enum_table
+
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
 
 statement error pgcode 0A000 creating partial statistics at extremes on bool and enum columns is disabled
 CREATE STATISTICS enum_table_full ON a FROM enum_table USING EXTREMES

--- a/pkg/sql/opt/exec/execbuilder/testdata/partial_stats
+++ b/pkg/sql/opt/exec/execbuilder/testdata/partial_stats
@@ -1255,6 +1255,11 @@ INSERT INTO ka VALUES (9999, NULL)
 statement ok
 CREATE STATISTICS ka_fullstat ON a FROM ka
 
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
+
 let $hist_id_ka_outer_buckets_full
 SELECT histogram_id FROM [SHOW STATISTICS FOR TABLE ka] WHERE statistics_name = 'ka_fullstat'
 
@@ -1323,6 +1328,11 @@ INSERT INTO ka VALUES (10002, 10001)
 statement ok
 CREATE STATISTICS ka_fullstat ON a FROM ka
 
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
+
 statement ok
 INSERT INTO ka VALUES (10003, 10002), (10004, NULL)
 
@@ -1356,13 +1366,18 @@ INSERT INTO ka SELECT x, x FROM generate_series(0, 9) as g(x)
 statement ok
 CREATE STATISTICS ka_fullstat ON a FROM ka
 
+# Clear the stat cache so that creating partial statistics has access to the
+# latest full statistic.
+statement ok
+SELECT crdb_internal.clear_table_stats_cache();
+
 statement ok
 INSERT INTO ka VALUES (10, 10)
 
 statement ok
 CREATE STATISTICS ka_partialstat ON a FROM ka USING EXTREMES
 
-query T
+query T retry
 EXPLAIN SELECT * FROM ka WHERE a > 5
 ----
 distribution: local

